### PR TITLE
Node shapes part 1, build time optimization, lazy WASM loader.

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description
- Implements https://github.com/luna/ide/issues/163
    - The application is now distributed to WASM binary (gzipped, not processed by webpack) and JS helpers bundle (gzipped, processed by webpack). 
    - There is a new loader implementation that starts downloading the bundles and as soon as it finishes downloading the helper bundle, it starts online WASM compilation (before the WASM binary finishes downloading).
    - The loader displays a nice progress bar to the user.
    - **The top-level `examples` folder was renamed to `app` and changed significantly. It is meant now to be the base for our cloud and native app. When you run it, please navigate to `http://localhost:8080/debug` to get the list of debug demo scenes.**
- Implements https://github.com/luna/ide/issues/164
    - Please use `./scripts/watch.sh --dev`. After changing any example scene, the time between hitting `ctrl+s` and seeing new results in the browser dropped from 45s to 3s.
- Implements https://github.com/luna/ide/issues/149

### Important Notes
**Please read the above points carefully!**

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/core/src/display/shape/primitive/glsl/color.glsl [hunk 1]
```diff
@@ -189,28 +235,164 @@ HSV hsv(RGB rgb) {
...
+        return sample(GRADIENT2(gradient.control_point3,gradient.control_point4),offset);          \
+    } else if (offset < gradient.control_point5.offset) {                                          \
+        return sample(GRADIENT2(gradient.control_point4,gradient.control_point5),offset);          \
+    } else {                                                                                       \
+        return gradient.control_point5.color;                                                      \
+    }                                                                                              \
+}                                                                                                  \
+
+
+DEF_GRADIENT(Rgb,rgb,RgbGradientControlPoint,RgbGradient1,RgbGradient2,RgbGradient3,RgbGradient4,RgbGradient5)
```
Comment (farmaazon) [inline lib/core/src/display/shape/primitive/glsl/color.glsl]: In c you can do concatenating identifiers in macros like"
`struct GRADIENT ## 1`
So you could have then
`DEF_GRADIENT(Rgb,rgb,RgbGradientControlPoint,RgbGradient)`
Maybe glsl also have such utility?

File: lib/core/src/display/shape/primitive/glsl/fragment_runner.glsl [hunk 2]
```diff
@@ -3,15 +3,14 @@
 Env   env      = Env(1);
 vec2  position = input_local.xy ;
 Shape shape    = run(env,position);
-float alpha    = render(shape);
+// float alpha    = render(shape);
```
Comment (farmaazon) [inline lib/core/src/display/shape/primitive/glsl/fragment_runner.glsl]: commented code

File: lib/core/src/display/shape/primitive/glsl/shape.glsl [hunk 3]
```diff
@@ -112,29 +112,70 @@ Id new_id_layer (BoundSdf sdf, int i) {
 }
 
 
+// Premultiplied
+struct Color {
```
Comment (farmaazon) [inline lib/core/src/display/shape/primitive/glsl/shape.glsl]: I wonder if "nice docs rule" apply also to glsl code? If yes, then this is certainly not nice doc.

You could also add docs to the other structures, although they can be also treated as "private API"

File: lib/core/src/display/symbol/shader.rs [hunk 4]
```diff
@@ -133,6 +133,7 @@ impl {
                 let shader      = shader_builder.build();
                 let vert_shader = compile_vertex_shader  (&self.context,&shader.vertex);
                 let frag_shader = compile_fragment_shader(&self.context,&shader.fragment);
+                println!("{}",&shader.fragment);
```
Comment (farmaazon) [inline lib/core/src/display/symbol/shader.rs]: to remove

File: lib/core/src/system/gpu/data/buffer/item.rs [hunk 5]
```diff
@@ -27,32 +27,42 @@ pub trait MatrixCtx<T,R,C> = where
 
 
 // ==================
-// === BufferItem ===
+// === Item ===
```
Comment (farmaazon) [inline lib/core/src/system/gpu/data/buffer/item.rs]: Adjust the boundaries length.

File: lib/gui/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  'WebGlVertexArrayObject',
+  'Window',
+  'console',
+  'EventTarget',
+  'Event',
+  'MouseEvent',
+  'WheelEvent',
+  'DomRect',
+  'AddEventListenerOptions'
+]
```
Comment (farmaazon) [inline lib/gui/Cargo.toml]:
```
                             .-----.
                            /7  .  (
                           /   .-.  \
                          /   /   \  \
                         / `  )   (   )
                        / `   )   ).  \
                      .'  _.   \_/  . |
     .--.           .' _.' )`.        |
    (    `---...._.'   `---.'_)    ..  \
     \            `----....___    `. \  |
      `.           _ ----- _   `._  )/  |
        `.       /"  \   /"  \`.  `._   |
          `.    ((O)` ) ((O)` ) `.   `._\
            `-- '`---'   `---' )  `.    `-.
               /                  ` \      `-.
             .'                      `.       `.
            /                     `  ` `.       `-.
     .--.   \ ===._____.======. `    `   `. .___.--`     .''''.
    ' .` `-. `.                )`. `   ` ` \          .' . '  8)
   (8  .  ` `-.`.               ( .  ` `  .`\      .'  '    ' /
    \  `. `    `-.               ) ` .   ` ` \  .'   ' .  '  /
     \ ` `.  ` . \`.    .--.     |  ` ) `   .``/   '  // .  /
      `.  ``. .   \ \   .-- `.  (  ` /_   ` . / ' .  '/   .'
        `. ` \  `  \ \  '-.   `-'  .'  `-.  `   .  .'/  .'
          \ `.`.  ` \ \    ) /`._.`       `.  ` .  .'  /
    LGB    |  `.`. . \ \  (.'               `.   .'  .'
        __/  .. \ \ ` ) \                     \.' .. \__
 .-._.-'     '"  ) .-'   `.                   (  '"     `-._.--.
(_________.-====' / .' /\_)`--..__________..-- `====-. _________)
```

File: lib/gui/src/easing_animator.rs [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
-use crate::system::web::create_element;
-use crate::system::web::NodeInserter;
-use crate::system::web::AttributeSetter;
-use crate::system::web::StyleSetter;
+use basegl::animation::easing::*;
+use basegl::animation::animator::easing::EasingAnimator;
+use basegl::system::web::create_element;
+use basegl::system::web::NodeInserter;
+use basegl::system::web::AttributeSetter;
+use basegl::system::web::StyleSetter;
```
Comment (farmaazon) [inline lib/gui/src/easing_animator.rs]: These imports should be reorganized.

File: lib/gui/src/glyph_system.rs [hunk 8]
```diff
@@ -1,10 +1,11 @@
...
-use crate::display::object::DisplayObject;
-use crate::display::shape::text::glyph::font::FontRegistry;
-use crate::display::shape::text::glyph::system::GlyphSystem;
-use crate::display::world::*;
+use basegl::display::object::DisplayObjectOps;
+use basegl::display::object::DisplayObject;
+use basegl::display::shape::text::glyph::font::FontRegistry;
+use basegl::display::shape::text::glyph::system::GlyphSystem;
+use basegl::display::world::*;
+use basegl::system::web;
```
Comment (farmaazon) [inline lib/gui/src/glyph_system.rs]: reorganize imports - this block is now the same category as the next one.

File: lib/gui/src/shapes.rs [hunk 9]
```diff
@@ -1,27 +1,27 @@
...
+use basegl::system::web::set_stack_trace_limit;
+use basegl::system::web::forward_panic_hook_to_console;
 
 use nalgebra::Vector2;
 use wasm_bindgen::prelude::*;
 
-use crate::display::shape::primitive::def::*;
-use crate::display::navigation::navigator::Navigator;
+use basegl::display::shape::primitive::def::*;
+//use basegl::display::navigation::navigator::Navigator;
```
Comment (farmaazon) [inline lib/gui/src/shapes.rs]: why it is commented? Remove or leave a todo

File: lib/gui/src/shapes.rs [hunk 10]
```diff
@@ -30,22 +30,34 @@ pub fn run_example_shapes() {
...
 
 fn init(world: &World) {
     let scene  = world.scene();
     let camera = scene.camera();
     let screen = camera.screen();
 
-    let navigator = Navigator::new(&scene, &camera);
-    let navigator = navigator.expect("Couldn't create navigator");
+//    let navigator = Navigator::new(&scene, &camera);
+//    let navigator = navigator.expect("Couldn't create navigator");
```
Comment (farmaazon) [inline lib/gui/src/shapes.rs]: ditto

File: lib/gui/src/shapes.rs [hunk 11]
```diff
@@ -108,9 +128,9 @@ pub fn frp_test (callback: Box<dyn Fn(f32,f32)>) -> MouseManager {
 
     // final_position.event.display_graphviz();
 
-    trace("X" , &debug.event);
+//    trace("X" , &debug.event);
 
-    final_position.map("foo",move|p| {callback(p.x as f32,-p.y as f32)});
+//    final_position.map("foo",move|p| {callback(p.x as f32,-p.y as f32)});
```
Comment (farmaazon) [inline lib/gui/src/shapes.rs]: ditto

File: lib/gui/src/sprite_system.rs [hunk 12]
```diff
@@ -1,27 +1,28 @@
...
-use crate::system::web::set_stdout;
+use basegl::display::object::DisplayObject;
+use basegl::display::object::DisplayObjectOps;
+use basegl::display::symbol::geometry::Sprite;
+use basegl::display::symbol::geometry::SpriteSystem;
+use basegl::display::world::*;
+use basegl::prelude::*;
+use basegl::system::web::forward_panic_hook_to_console;
+use basegl::system::web::set_stdout;
+use basegl::system::web;
```
Comment (farmaazon) [inline lib/gui/src/sprite_system.rs]: imports again

File: lib/gui/src/text_selecting.rs [hunk 13]
```diff
@@ -2,14 +2,14 @@
...
-use crate::system::web;
-use crate::system::web::forward_panic_hook_to_console;
+use basegl::display::world::WorldData;
+use basegl::display::object::DisplayObjectOps;
+use basegl::display::shape::text::glyph::font::FontRegistry;
+use basegl::display::shape::text::text_field::TextField;
+use basegl::display::shape::text::text_field::TextFieldProperties;
+use basegl::display::world::*;
+use basegl::system::web;
+use basegl::system::web::forward_panic_hook_to_console;
```
Comment (farmaazon) [inline lib/gui/src/text_selecting.rs]: And this is the last "reorganize imports plz" place I mark with comment. Please look through all files in this crate

File: lib/gui/src/text_typing.rs [hunk 14]
```diff
@@ -39,7 +40,7 @@ pub fn run_example_text_typing() {
         let now             = js_sys::Date::now();
         let animation_start = now + 3000.0;
         let start_scrolling = animation_start + 10000.0;
-        let mut chars       = typed_character_list(animation_start,include_str!("../lib.rs"));
+        let mut chars       = typed_character_list(animation_start,include_str!("../../core/src/lib.rs"));
```
Comment (farmaazon) [inline lib/gui/src/text_typing.rs]: You can pick any other file as well :D

File: lib/core/src/display/shape/primitive/glsl/color.glsl (hunk 1 above)
Comment (wdanilo) [reply lib/core/src/display/shape/primitive/glsl/color.glsl]: It has, but not in WebGL -.-

File: lib/core/src/data/color.rs [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+// === Gradient ===
+// ================
+
+/// A linear color gradient implementation. It accepts control points which base on colors for
+/// a given palette. The palette need to define `mix` operation on GPU.
+#[derive(Clone,Debug,Derivative)]
+#[derivative(Default(bound=""))]
+pub struct Gradient<Color> {
+    control_points : Vec<(f32,Color)>,
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/data/color.rs]: Fixed!

File: lib/core/src/data/color.rs [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    /// Constructor.
+    pub fn new(gradient:Gradient) -> Self {
+        let min_distance = 0.0;
+        let max_distance = 10.0;
+        let slope        = Slope::Smooth;
+        Self {min_distance,max_distance,slope,gradient}
+    }
+
+    /// Setter for the `max_distance` field.
+    pub fn max_distance(mut self, t:f32) -> Self {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/data/color.rs]: Its a constructor setter. Fixed docs

File: lib/core/src/display/shape/primitive/glsl/shape.glsl (hunk 3 above)
Comment (wdanilo) [reply lib/core/src/display/shape/primitive/glsl/shape.glsl]: For now its private API. But I want to make docs nicer there in general

File: lib/core/src/display/symbol/shader.rs (hunk 4 above)
Comment (wdanilo) [reply lib/core/src/display/symbol/shader.rs]: ooops, thanks for catching.

File: lib/gui/src/easing_animator.rs (hunk 7 above)
Comment (wdanilo) [reply lib/gui/src/easing_animator.rs]: Ok, doing it, but it was not my code originally. CC @notdanilo  - take care about sorting imports please.

File: lib/gui/src/shapes.rs (hunk 9 above)
Comment (wdanilo) [reply lib/gui/src/shapes.rs]: Adding notice to the top of the file as we agreed.

**Pull request merged**